docs: update README with SingleStore changes (#17)
memsql/go-singlestore-driver: README.md

# Go-SingleStore-Driver

> A SingleStore-Driver for Go's [database/sql](https://golang.org/pkg/database/sql/) package.

Reviewer: ```suggestion
A driver for Go's [database/sql](https://golang.org/pkg/database/sql/) package provided by SingleStore.
```

  * Optional placeholder interpolation
  * Supports zlib compression.

## Changes from Upstream Go-MySQL-Driver

This driver is a fork of [go-sql-driver/mysql](https://github.com/go-sql-driver/mysql) with the following SingleStore-specific changes:

### New Features

* **Active Query Cancellation** (v1.9.2-p1): When a context is cancelled during query execution, if `ctxCancellationEnabled` is set to `true`, the driver now opens a new connection and executes `KILL QUERY <connection_id> <aggregator_id>` to terminate the running query on the server, providing immediate resource cleanup. See [Query Cancellation](#query-cancellation) for details.

> * **`ctxCancellationEnabled` DSN Parameter** (v1.9.2-p2): A boolean flag to enable/disable the active query cancellation feature. It is set to `false` by default for backward compatibility, but is recommended to set this to `true` for SingleStore connections.

Reviewer: ```suggestion
* **`ctxCancellationEnabled` DSN Parameter** (v1.9.2-p2): A boolean flag to enable/disable the active query cancellation feature. It is set to `false` by default for backward compatibility, but is recommended to set this to `true` for SingleStore connections. Refer to [ctxCancellationEnabled](#ctxCancellationEnabled) for more information.
```

* **`skipParseNumbers` DSN Parameter** (v1.8.1-p2): When set to `true`, tells the driver not to convert integers and floats to native Go data types, passing their string representation as returned by the server to `database/sql`. See [`skipParseNumbers`](#skipparsenumbers) for details.

### Other Changes

* **Client Name**: The `_client_name` connection attribute is set to `memsql/go-singlestore-driver` to distinguish from the upstream driver.

* **SingleStore Compatibility Notes**: Some MySQL-specific parameters like `charset`, `collation`, `allowOldPasswords`, `rejectReadOnly`, and `serverPubKey` have no effect on SingleStore. These are documented in the [Parameters](#parameters) section.

## Requirements

* Go 1.21 or higher. We aim to support the 3 latest versions of Go.
* SingleStore 8.7 or higher.

---------------------------------------

## Installation

> Simply install the package to your [$GOPATH](https://github.com/golang/go/wiki/GOPATH "GOPATH") with the [go tool](https://golang.org/cmd/go/ "go command") from shell:

Reviewer: ```suggestion
Install the package to your [$GOPATH](https://github.com/golang/go/wiki/GOPATH "GOPATH") with the [go tool](https://golang.org/cmd/go/ "go command") from shell:
```

```bash
go get -u github.com/memsql/go-singlestore-driver
```
Make sure [Git is installed](https://git-scm.com/downloads) on your machine and in your system's `PATH`.

## Usage

> _Go SingleStore Driver_ is an implementation of Go's `database/sql/driver` interface. You only need to import the driver and can use the full [`database/sql`](https://golang.org/pkg/database/sql/) API then.

Reviewer: ```suggestion
_Go SingleStore Driver_ is an implementation of Go's `database/sql/driver` interface. Once you import the driver, you can access the full [`database/sql`](https://golang.org/pkg/database/sql/) API.
```

> Use `mysql` as `driverName` and a valid [DSN](#dsn-data-source-name) as `dataSourceName`:

Reviewer: ```suggestion
Use `singlestore` as `driverName` and a valid [DSN](#dsn-data-source-name) as `dataSourceName`:
```

Reviewer: Should this be `singlestore` instead of `mysql`?

Author: not yet, the `driverName` is still `mysql` for backward compatibility. At some point we may rename it to `singlestore`, but not yet.

Valid Values:   true, false
Default:        false
```
**Warning**: this setting has no effect for SingleStore.

> `allowOldPasswords=true` allows the usage of the insecure old password method. This should be avoided, but is necessary in some cases. See also [the old_passwords wiki page](https://github.com/go-sql-driver/mysql/wiki/old_passwords).

Reviewer: ```suggestion
`allowOldPasswords=true` allows the usage of the insecure old password method. In SingleStore, this has no effect and exists only for compatibility.
```

Default:        none
```

**Warning**: this setting has no effect for SingleStore.

> Sets the charset used for client-server interaction (`"SET NAMES <value>"`). If multiple charsets are set (separated by a comma), the following charset is used if setting the charset fails. This enables for example support for `utf8mb4` ([introduced in MySQL 5.5.3](http://dev.mysql.com/doc/refman/5.5/en/charset-unicode-utf8mb4.html)) with fallback to `utf8` for older servers (`charset=utf8mb4,utf8`).

Reviewer: ```suggestion
Sets the charset used for client-server interaction (`"SET NAMES <value>"`). In SingleStore, this has no effect and exists only for compatibility.
```

Reviewer: Because this is no-op in SingleStore, I recommend, not having a description for this.

Default:        false
```

> `tls=true` enables TLS / SSL encrypted connection to the server. Use `skip-verify` if you want to use a self-signed or invalid certificate (server side) or use `preferred` to use TLS only when advertised by the server. This is similar to `skip-verify`, but additionally allows a fallback to a connection which is not encrypted. Neither `skip-verify` nor `preferred` add any reliable security. You can use a custom TLS config after registering it with [`mysql.RegisterTLSConfig`](https://godoc.org/github.com/memsql/go-singlestore-driver#RegisterTLSConfig).

Reviewer: singlestore.RegisterTLSConfig?

### `LOAD DATA LOCAL INFILE` support
For this feature you need direct access to the package. Therefore you must change the import path (no `_`):
```go
import "github.com/memsql/go-singlestore-driver"
```

Files must be explicitly allowed by registering them with `mysql.RegisterLocalFile(filepath)` (recommended) or the allowlist check must be deactivated by using the DSN parameter `allowAllFiles=true` ([*Might be insecure!*](https://dev.mysql.com/doc/refman/8.0/en/load-data.html#load-data-local)).

To use a `io.Reader` a handler function must be registered with `mysql.RegisterReaderHandler(name, handler)` which returns a `io.Reader` or `io.ReadCloser`. The Reader is available with the filepath `Reader::<name>` then. Choose different names for different handlers and `DeregisterReaderHandler` when you don't need it anymore.

See the [godoc of Go-SingleStore-Driver](https://godoc.org/github.com/memsql/go-singlestore-driver "golang singlestore driver documentation") for details.

### `time.Time` support
The default internal output type of `DATE` and `DATETIME` values is `[]byte` which allows you to scan the value into a `[]byte`, `string` or `sql.RawBytes` variable in your program.

> However, many want to scan `DATE` and `DATETIME` values into `time.Time` variables, which is the logical equivalent in Go to `DATE` and `DATETIME` in SQL. You can do that by changing the internal output type from `[]byte` to `time.Time` with the DSN parameter `parseTime=true`. You can set the default [`time.Time` location](https://golang.org/pkg/time/#Location) with the `loc` DSN parameter.

Reviewer: ```suggestion
However, many programs want to scan `DATE` and `DATETIME` values into `time.Time` variables, which is the logical equivalent in Go to `DATE` and `DATETIME` in SingleStore. You can do that by changing the internal output type from `[]byte` to `time.Time` with the DSN parameter `parseTime=true`. You can set the default [`time.Time` location](https://golang.org/pkg/time/#Location) with the `loc` DSN parameter.
```
